### PR TITLE
 Ability to filter threads by state

### PR DESCRIPTION
**Pull request by v-nikulichev**
The new syntax is tbstack <pid>/<state list>. For example,
```$ tbstack 1234/R```
Will show only threads in state Running.

### CODE REVIEW

File: src/proc.c [hunk 1]
```diff
@@ -290,6 +290,26 @@ int adjust_threads(int *tids, int nr_tids, int *user_tids,
     return 0;
 }
 
+void filter_threads(int *tids, int *index, char *states, int *nr_tids,
```
Comment (dmitri-shubin) [inline src/proc.c]: Take `nr_tids` by value and return new count as a result?

Comment (dmitri-shubin) [reply src/proc.c]: And what about using array syntax for arguments that are ... well ... arrays? :)

File: src/tbstack.c [hunk 2]
```diff
@@ -90,6 +92,22 @@ static void parse_pid_arg(const char *prog, char *arg)
     if (tstr != NULL) {
         char c, prev = '\0';
         int i = 0, j;
+        int is_state_list;
```
Comment (dmitri-shubin) [inline src/tbstack.c]: Why not init it right here?

File: src/tbstack.c [hunk 3]
```diff
@@ -90,6 +92,22 @@ static void parse_pid_arg(const char *prog, char *arg)
     if (tstr != NULL) {
```
Comment (dmitri-shubin) [inline src/tbstack.c]: BTW I'd return here if `tstr` is `NULL` to save indent level :)

Or move list parsing to separate function.

File: src/proc.c (hunk 1 above)
Comment (v-nikulichev) [reply src/proc.c]: Fixed

File: src/tbstack.c (hunk 2 above)
Comment (v-nikulichev) [reply src/tbstack.c]: Fixed

File: src/tbstack.c (hunk 3 above)
Comment (v-nikulichev) [reply src/tbstack.c]: Fixed in separate commit: https://github.com/tbricks/tbstack/pull/24/commits/c120509ef6df7130ba6ebc4caa367eb1fd70031f

**Comment by v-nikulichev** [comment]
@dmitri-shubin Added fixes and squashed it with original commit (+one new for refactoring).

File: src/proc.c [hunk 4]
```diff
@@ -290,6 +290,27 @@ int adjust_threads(int *tids, int nr_tids, int *user_tids,
     return 0;
 }
 
+int filter_threads(int tids[], int index[], char states[], int nr_tids,
+        const char user_states[])
```
Comment (dmitri-shubin) [inline src/proc.c]: `const char *user_states` since it's actually NUL-terminated _string_.

Comment (v-nikulichev) [reply src/proc.c]: Fixed

**Comment by dmitri-shubin** [comment]
Merged!
Thanks!

**Pull request merged**